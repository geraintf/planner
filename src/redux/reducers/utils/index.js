import uuidv4 from 'uuid/v4';
import { arrayMove } from 'react-sortable-hoc';

import { getDayTodos } from '../../selectors';

import { syncDayTodos } from '../../../services/requests';

export const toggleTodos = (state, todoId) => {
  const toggle = (todos,id) => todos.map(todo => (
    todo.todoId === id
      ? { ...todo, completed: !todo.completed }
      : todo
  ));

  const updatedTodos = {
    ...state.todos,
    [state.selectedDate]: toggle(getDayTodos(state), todoId)
  };

  syncDayTodos(state.selectedDate, updatedTodos[state.selectedDate]);

  return {
    ...state,
    todos: updatedTodos
  };
};

export const addTodo = (state, text) => {
  const newTodo = {
    todoId: uuidv4(),
    text,
    completed: false
  };

  const updatedTodos = {
    ...state.todos,
    [state.selectedDate]: [...getDayTodos(state)].concat([newTodo])
  };

  syncDayTodos(state.selectedDate, updatedTodos[state.selectedDate]);

  return {
    ...state,
    todos: updatedTodos
  };
};


export const removeTodo = (state, todoId) => {
  const updatedTodos = {
    ...state.todos,
    [state.selectedDate]: getDayTodos(state).filter(todo => todo.todoId !== todoId)
  };

  syncDayTodos(state.selectedDate, updatedTodos[state.selectedDate]);

  return {
    ...state,
    todos: updatedTodos
  };
};

export const moveTodo = (state, oldIndex, newIndex) => {
  const updatedTodos = {
    ...state.todos,
    [state.selectedDate]: arrayMove(getDayTodos(state), oldIndex, newIndex)
  };

  syncDayTodos(state.selectedDate, updatedTodos[state.selectedDate]);

  return {
    ...state,
    todos: updatedTodos
  };
};

export const editTodo = (state, todoId, newValue) => {
  const edit = (todos, id, newValue) =>
    todos.map(todo => {
      if (todo.todoId === id) todo.text = newValue;

      return todo;
    });

  const updatedTodos = {
    ...state.todos,
    [state.selectedDate]: edit(getDayTodos(state), todoId, newValue)
  };

  syncDayTodos(state.selectedDate, updatedTodos[state.selectedDate]);

  return {
    ...state,
    todos: updatedTodos
  };
};
