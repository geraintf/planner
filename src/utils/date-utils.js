import { format } from 'date-fns';

const DATE_FORMAT = 'DD-MM-YYYY';

const formatTodoDateKey = date => format(date, DATE_FORMAT).toString();

export {
  formatTodoDateKey
};
