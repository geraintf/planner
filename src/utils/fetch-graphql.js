require('isomorphic-fetch');

const opts = {
  method: 'POST',
  headers: { 'Content-Type': 'application/json' },
  credentials: 'same-origin'
};

export const makeGraphqlReq = ({ query, operationName, variables }, path = '') => {

  if(!query) {
    throw new Error('An invalid query was passed into makeGraphqlReq: ' + query);
  }

  return fetch(`${path}/graphql`, {
    ...opts,
    body: JSON.stringify({ query, operationName, variables })
  }).then(res => res.json());
};
