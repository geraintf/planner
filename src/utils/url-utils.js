
export const getFullUrl = req => req.protocol + '://' + req.get('host');
