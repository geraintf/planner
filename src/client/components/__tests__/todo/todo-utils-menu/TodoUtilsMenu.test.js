import React from 'react';
import { shallow } from 'enzyme';

import { unwrapComponent } from '../../../../../../test/utils/unwrap';

import WrappedTodoUtilsMenu from '../../../todo/todo-utils-menu/TodoUtilsMenu';
const TodoUtilsMenu = unwrapComponent(WrappedTodoUtilsMenu);

describe('TodoUtilsMenu', () => {
  let output;
  let props;

  beforeEach(() => {
    props = {
      removeTodo: jest.fn(() => {}),
      toggleEditing: jest.fn(() => {}),
      todoId: '123456789'
    };
    output = shallow(<TodoUtilsMenu {...props} />);
  });

  it('should render correctly', () => {
    expect(output).toMatchSnapshot();
  });

  it('should render a dropdown toggle and dropdown menu', () => {
    expect(
      output
        .find('DropdownToggle')
        .length
    ).toEqual(1);

    expect(
      output
        .find('DropdownMenu')
        .length
    ).toEqual(1);
  });

  it('should two menu items', () => {
    expect(
      output
        .find('.todo-utils-menu__item')
        .length
    ).toEqual(2);
  });

  it('should toggle the open state of the dropdown', () => {
    output
      .find('DropdownToggle')
      .simulate('click');

    expect(output.state('open')).toEqual(true);
  });

  it('should handle the removing of a todo', () => {
    output
      .find('.remove-item')
      .simulate('click');

    expect(output.state('open')).toEqual(true);
    expect(props.removeTodo).toHaveBeenCalled();
  });

  it('should handle the edit toggling', () => {
    output
      .find('.edit-item')
      .simulate('click');

    expect(output.state('open')).toEqual(true);
    expect(props.toggleEditing).toHaveBeenCalled();
  });
});
