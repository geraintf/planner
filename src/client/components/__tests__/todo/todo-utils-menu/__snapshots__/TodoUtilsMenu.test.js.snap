// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoUtilsMenu should render correctly 1`] = `
<Dropdown
  active={false}
  addonType={false}
  className="todo-utils-menu"
  direction="down"
  inNavbar={false}
  isOpen={false}
  nav={false}
  toggle={[Function]}
>
  <DropdownToggle
    aria-expanded={false}
    aria-haspopup={true}
    color="secondary"
    data-toggle="dropdown"
    onClick={[Function]}
    tag="a"
  >
    <FontAwesomeIcon$1
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            320,
            512,
            Array [],
            "f107",
            "M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z",
          ],
          "iconName": "angle-down",
          "prefix": "fas",
        }
      }
      listItem={false}
      mask={null}
      name=""
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      transform={null}
    />
  </DropdownToggle>
  <DropdownMenu
    flip={true}
    right={true}
    tag="div"
  >
    <div
      className="todo-utils-menu__item edit-item"
      onClick={[Function]}
    >
      <div
        className="todo-utils-menu__item-icon"
      >
        <FontAwesomeIcon$1
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                576,
                512,
                Array [],
                "f044",
                "M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z",
              ],
              "iconName": "edit",
              "prefix": "fas",
            }
          }
          listItem={false}
          mask={null}
          name=""
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          symbol={false}
          transform={null}
        />
      </div>
      <div
        className="todo-utils-menu__item-text"
      >
        Edit
      </div>
    </div>
    <div
      className="todo-utils-menu__item remove-item"
      onClick={[Function]}
    >
      <div
        className="todo-utils-menu__item-icon"
      >
        <FontAwesomeIcon$1
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                384,
                512,
                Array [],
                "f00d",
                "M323.1 441l53.9-53.9c9.4-9.4 9.4-24.5 0-33.9L279.8 256l97.2-97.2c9.4-9.4 9.4-24.5 0-33.9L323.1 71c-9.4-9.4-24.5-9.4-33.9 0L192 168.2 94.8 71c-9.4-9.4-24.5-9.4-33.9 0L7 124.9c-9.4 9.4-9.4 24.5 0 33.9l97.2 97.2L7 353.2c-9.4 9.4-9.4 24.5 0 33.9L60.9 441c9.4 9.4 24.5 9.4 33.9 0l97.2-97.2 97.2 97.2c9.3 9.3 24.5 9.3 33.9 0z",
              ],
              "iconName": "times",
              "prefix": "fas",
            }
          }
          listItem={false}
          mask={null}
          name=""
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          symbol={false}
          transform={null}
        />
      </div>
      <div
        className="todo-utils-menu__item-text"
      >
        Delete
      </div>
    </div>
  </DropdownMenu>
</Dropdown>
`;
